File: VB/WindowsApplication78/MyColorEdit.vb
Imports System
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Registrator
Imports System.Drawing
Imports System.Collections

Namespace CustomEditors

    <UserRepositoryItem("Register")>
    Public Class RepositoryItemMyColorEdit
        Inherits RepositoryItemColorEdit

        Shared Sub New()
            Call Register()
        End Sub

        Public Sub New()
        End Sub

        Friend Const EditorName As String = "MyColorEdit"

        Public Shared Sub Register()
            Call EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(EditorName, GetType(MyColorEdit), GetType(RepositoryItemMyColorEdit), GetType(ViewInfo.ColorEditViewInfo), New DevExpress.XtraEditors.Drawing.ColorEditPainter(), True, CType(Nothing, System.Drawing.Image)))
        End Sub

        Public Overrides ReadOnly Property EditorTypeName As String
            Get
                Return EditorName
            End Get
        End Property

        Protected Overrides Function GetColorDisplayText(ByVal editValue As Color) As String
            Dim color As Object = editValue
            ColorHelper.TryConvertToKnownColor(color)
            Return MyBase.GetColorDisplayText(CType(color, Color))
        End Function
    End Class

    Friend Class ColorHelper

        Shared Sub New()
            knownColorsHash = InitKnownColorsHash()
        End Sub

        Private Sub New()
        End Sub

        Private Shared knownColorsHash As Hashtable

        Private Shared Function InitKnownColorsHash() As Hashtable
            Dim knownColors As Array = [Enum].GetValues(GetType(KnownColor))
            Dim hashtable As Hashtable = New Hashtable(knownColors.Length)
            For Each knownColorObject As KnownColor In knownColors
                Dim color As Color = Color.FromKnownColor(knownColorObject)
                Dim argbValue As Integer = color.ToArgb()
                If color.IsSystemColor AndAlso hashtable.ContainsKey(argbValue) Then Continue For
                hashtable(argbValue) = knownColorObject
            Next

            Return hashtable
        End Function

        Public Shared Function TryConvertToKnownColor(ByRef color As Object) As Boolean
            Dim knownColor As KnownColor = GetKnownColor(color)
            If knownColor <> 0 Then
                color = Drawing.Color.FromKnownColor(knownColor)
                Return True
            End If

            Return False
        End Function

        Public Shared Function GetKnownColor(ByVal color As Color) As KnownColor
            Dim knownColor As KnownColor = color.ToKnownColor()
            If knownColor = 0 Then knownColor = GetKnownColor(color.ToArgb())
            Return knownColor
        End Function

        Public Shared Function GetKnownColor(ByVal argb As Integer) As KnownColor
            Dim result As Object = knownColorsHash(argb)
            If result Is Nothing Then Return 0
            Return CType(result, KnownColor)
        End Function

        Public Shared Function GetKnownColor(ByVal color As Object) As KnownColor
            If TypeOf color Is KnownColor Then Return CType(color, KnownColor)
            If TypeOf color Is Color Then Return GetKnownColor(CType(color, Color))
            If TypeOf color Is Integer Then Return GetKnownColor(CInt(color))
            Return 0
        End Function
    End Class

    Public Class MyColorEdit
        Inherits ColorEdit

        Shared Sub New()
            RepositoryItemMyColorEdit.Register()
        End Sub

        Public Sub New()
        End Sub

        Public Overrides ReadOnly Property EditorTypeName As String
            Get
                Return RepositoryItemMyColorEdit.EditorName
            End Get
        End Property
    End Class
End Namespace

File: CS/WindowsApplication78/MyColorEdit.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Popup;
using System.Drawing;
using System.Collections;

namespace CustomEditors
{
    [UserRepositoryItem("Register")]
    public class RepositoryItemMyColorEdit : RepositoryItemColorEdit
    {
        static RepositoryItemMyColorEdit()
        {
            Register();
        }
        public RepositoryItemMyColorEdit() { }

        internal const string EditorName = "MyColorEdit";

        public static void Register()
        {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(EditorName, typeof(MyColorEdit),
                typeof(RepositoryItemMyColorEdit), typeof(DevExpress.XtraEditors.ViewInfo.ColorEditViewInfo),
                new DevExpress.XtraEditors.Drawing.ColorEditPainter(), true, null));
        }
        public override string EditorTypeName
        {
            get { return EditorName; }
        }

        protected override string GetColorDisplayText(Color editValue)
        {
            object color = editValue;
            ColorHelper.TryConvertToKnownColor(ref color);
            return base.GetColorDisplayText((Color)color);
        }
    }

    class ColorHelper
    {
        static ColorHelper()
        {
            knownColorsHash = InitKnownColorsHash();
        }
        private ColorHelper() { }

        private static Hashtable knownColorsHash;

        private static Hashtable InitKnownColorsHash()
        {
            System.Array knownColors = System.Enum.GetValues(typeof(KnownColor));
            Hashtable hashtable = new Hashtable(knownColors.Length);
            foreach (KnownColor knownColorObject in knownColors)
            {
                Color color = Color.FromKnownColor(knownColorObject);
                int argbValue = color.ToArgb();
                if (color.IsSystemColor && hashtable.ContainsKey(argbValue))
                    continue;
                hashtable[argbValue] = knownColorObject;
            }
            return hashtable;
        }

        public static bool TryConvertToKnownColor(ref object color)
        {
            KnownColor knownColor = GetKnownColor(color);
            if (knownColor != 0)
            {
                color = Color.FromKnownColor(knownColor);
                return true;
            }
            return false;
        }

        public static KnownColor GetKnownColor(Color color)
        {
            KnownColor knownColor = color.ToKnownColor();
            if (knownColor == 0)
                knownColor = GetKnownColor(color.ToArgb());
            return knownColor;
        }
        public static KnownColor GetKnownColor(int argb)
        {
            object result = knownColorsHash[argb];
            if (result == null)
                return 0;
            return (KnownColor)result;
        }

        public static KnownColor GetKnownColor(object color)
        {
            if (color is KnownColor)
                return (KnownColor)color;
            if (color is Color)
                return GetKnownColor((Color)color);
            if (color is int)
                return GetKnownColor((int)color);
            return 0;
        }
    }

    public class MyColorEdit : ColorEdit
    {
        static MyColorEdit()
        {
            RepositoryItemMyColorEdit.Register();
        }
        public MyColorEdit() { }

        public override string EditorTypeName
        {
            get { return RepositoryItemMyColorEdit.EditorName; }
        }
    }
}

